### DISCUSSION TITLE
TypeScript definitions for window.chrome.webview

### DISCUSSION THREAD

**Discussion by danbahrami** [MicrosoftEdge/WebView2Feedback#4028] [Q&A]
Hi, I'm looking through the WebView2 docs and can't find any API reference for `window.chrome.webview`. Specifically I'm looking to define a global TypeScript interface for `window.chrome.webview.postMessage` and `window.chrome.webview.addEventListener`.

I have a couple of questions:

### 1. What's the method for removing event listeners?
I can't tell if `window.chrome.addEventListener` returns an unsubscribe callback or if there's an adjacent `removeEventListener` that behaves the same as adding/removing DOM events.

### 2. Do you have a recommendation for a global TypeScript definition?
Here's what I've got so far but I'm not sure if it's accurate

```ts
declare global {
    interface Window {
        chrome?: {
            webview?: {
                addEventListener: (
                    message: string,
                    handler: (event: any) => void, // not sure yet what the event type should be
                ) => void;
                postMessage: (message: string) => void;
            };
        };
    }
}
```

**Comment by david-risney** [comment]
It behaves like the DOM add/removeEventListener. You can subscribe to the `message` event and the `sharedbufferreceived` event via `window.chrome.webview.addEventListener` and unsubcribe `window.chrome.webview.removeEventListener`.

We've recently added [ref docs for the JS APIs we have in the WebView2](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/javascript/) which can hopefully answer your questions. The JS ref docs are new so let me know if you see problems. 

In the future I'd also like for us to publish a ts.d file somewhere. Maybe include it in our nuget?

**Comment by jeremyong** [comment]
FYI I made an attempt here to include it in the standard @types repo but there are a few issues with this attempt, namely the lack of proper proxy property handling (currently, I'm using a catch-all `[key: string]: any` indexer which isn't quite right I don't think (I'm not a TS expert):

https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...jeremyong:DefinitelyTyped:master

Anyone is welcome to work from here as a starting point though.

**Reply by RampantDespair** [reply]
@jeremyong 
What happened to this? I can't find it on the repository's main branch.

**Comment by david-risney** [comment]
Thanks for doing that!

**Comment by jeremyong** [comment]
> Thanks for doing that!

No problem! I should mention that I think there are a few fields missing in the docs I had to add locally. Specifically, the event listener is documented as being the same as the built-in event listener, but I believe WebView2 adds non-standard fields to the event passed to the handler. For example, when posting a shared buffer object, the event receives an `additionalData` property and `getBuffer: () => ArrayBuffer` property. There may be other cases also but I'm a relatively new user of the WebView2 API so you likely would know those corner cases best. *edit: I did add the `SharedBufferReceivedEvent` to the file but haven't modified the event listener types yet*

**Comment by BalassaMarton** [comment]
@jeremyong would you submit a PR? I would absolutely use your code, but we need proper licensing, which is covered if the code is part of `@types`

**Comment by jeremyong** [comment]
@BalassaMarton I haven't been using WebView2 recently and have been a bit swamped with other work so I don't have time to clean up the PR with tests needed for now. However, to unblock you:

> As the author of [this commit](https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...jeremyong:DefinitelyTyped:master), I hereby release all code therein into the public domain for anyone to use as they wish.

Hopefully that helps you move forward with your project, but let me know if something else is needed.

**Comment by bgever** [comment]
Thanks for your type definitions @jeremyong  and releasing it to the public.

I created these two helper functions based on those types (below code is public domain).

`getWebView` introduces nullability checks for when the page isn't hosted within a webview.
`getHostObject` removes the index from the host object, so you can strictly type it and get errors for mistyped interface names.

```ts
export const getWebView = (): WebView | undefined => window?.chrome?.webview

type RemoveIndex<T> = {
    [K in keyof T as {} extends Record<K, 1> ? never : K]: T[K]
}

export const getHostObject = <T extends HostObjectAsyncProxy>(name: string) =>
    getWebView()?.hostObjects[name] as RemoveIndex<T> | undefined
```

Usage:
```ts
interface MyHostObject extends HostObjectAsyncProxy {
    DoSomething: () => Promise<void>
}

getHostObject<MyHostObject>('my')?.DoSomething() // ok
getHostObject<MyHostObject>('my')?.Oops() // errors
```

**Comment by bgever** [comment]
This definition is missing in the repo:

```ts
export interface WebView extends EventTarget {
    /**
     * When the page calls `postMessageWithAdditionalObjects`, the `message` parameter is sent to WebView2 in the same
     * fashion as {@link WebView.postMessage}.
     * Objects passed as 'additionalObjects' are converted to their native types and will be available in the
     * `CoreWebView2WebMessageReceivedEventArgs.AdditionalObjects` property.
     * @param message The message to send to the WebView2 host. This can be any object that can be serialized to JSON.
     * @param additionalObjects A sequence of DOM objects that have native representations in WebView2.
     * This parameter needs to be ArrayLike.
     *
     * The following DOM types are mapped to native:
     * DOM    | Win32              | .NET                | WinRT
     * ------ | ------------------ | ------------------- | ---------------------------
     * File   | ICoreWebView2File  | System.IO.FileInfo  | Windows.Storage.StorageFile
     *
     * `null` or `undefined` entries will be passed as `null` type in WebView2. Otherwise if an invalid or unsupported
     * object is passed via this API, an exception will be thrown and the message will fail to post.
     */
    postMessageWithAdditionalObjects(message: any, additionalObjects: ArrayLike<any>): void;
}
```

**Comment by RampantDespair** [comment]
Hello there,

I just created my own iteration of WebView2 TypeScript types.

GitHub repository:
https://github.com/RampantDespair/webview2-types

npm package:
https://www.npmjs.com/package/webview2-types

Feel free to check it out or contribute if you find anything missing!
